test/controllers/admin/comfy/external_api_clients_controller_test.rb

```diff
@@ -38,27 +75,402 @@ class Comfy::Admin::ExternalApiClientsControllerTest < ActionDispatch::Integrati
+                  },
+        model_definition: "class ExternalApiModelExample; def initialize(parameters); # do init stuff; end; def start; return true; end; def log; return true; end; end; # at the end of the file we have to implicitly return the class; ExternalApiModelExample;"
+      }
+    }
+
+    @user.update(can_manage_api: false)
+    sign_in(@user)
+
+    assert_no_difference "ExternalApiClient.count" do
+      post api_namespace_external_api_clients_url(api_namespace_id: @api_namespace.id), params: payload
+    end
+
+    expected_message = "You do not have the permission to do that. Only users who can_manage_api are allowed to perform that action."
+    assert_match expected_message, flash[:alert]
+    assert_response :redirect
+  end
+
+  test "#show: denies if user not signed in" do
+    get api_namespace_external_api_client_url(api_namespace_id: @api_namespace.id, id: @external_api_client.id)
+    expected_message = "You need to sign in or sign up before continuing."
+
+    assert_response :redirect
+    assert_match expected_message, flash[:alert]
+  end
+
+  test "#show: denies if user not permissioned to manage api" do
+    @user.update(can_manage_api: false)
+    sign_in(@user)
+
+    get api_namespace_external_api_client_url(api_namespace_id: @api_namespace.id, id: @external_api_client.id)
+    expected_message = "You do not have the permission to do that. Only users who can_manage_api are allowed to perform that action."
+
+    assert_response :redirect
+    assert_match expected_message, flash[:alert]
+  end
+
+  test "#show: allows if permissioned user is signed in" do
+    @user.update(can_manage_api: true)
+    sign_in(@user)
+
+    get api_namespace_external_api_client_url(api_namespace_id: @api_namespace.id, id: @external_api_client.id)
+
+    assert_response :success
+  end
+
+  test "#edit: denies if user not signed in" do
+    get edit_api_namespace_external_api_client_url(api_namespace_id: @api_namespace.id, id: @external_api_client.id)
+    expected_message = "You need to sign in or sign up before continuing."
+
+    assert_response :redirect
+    assert_match expected_message, flash[:alert]
+  end
+
+  test "#edit: denies if user not permissioned to manage api" do
+    @user.update(can_manage_api: false)
+    sign_in(@user)
+
+    get edit_api_namespace_external_api_client_url(api_namespace_id: @api_namespace.id, id: @external_api_client.id)
+    expected_message = "You do not have the permission to do that. Only users who can_manage_api are allowed to perform that action."
+
+    assert_response :redirect
+    assert_match expected_message, flash[:alert]
+  end
+
+  test "#edit: allows if permissioned user is signed in" do
+    @user.update(can_manage_api: true)
+    sign_in(@user)
+
+    get edit_api_namespace_external_api_client_url(api_namespace_id: @api_namespace.id, id: @external_api_client.id)
+
+    assert_response :success
+  end
+
+  test "#update: should allow when permissioned user is signed-in" do
```

Comment:
@alis-khadka can you add some assertions to this test that validates passed parameters are saved correctly on the DB (eg: metadata hash is saved the with the same keys/values, model_definition is correct etc)